Add a tool for measuring expectation values of Pauli strings with readout error mitigation

ddddddanni opened the pull request:
This PR does the following two things:

1. Improved Readout Benchmarking: Modified run_shuffled_with_readout_benchmarking to accept a list of qubit tuples (List[Tuple[Qubit]]). This allows for calculating the readout error rate on specific qubit subsets, ensuring alignment with the qubit sets used in non-identity Pauli operators within Pauli strings.

2. Added a tool to compute expectation values for Pauli operators. This tool calculates both unmitigated and readout-error-mitigated expectation values for each circuit, and integrates the results from the readout benchmarking.

ddddddanni commented:
Hey @eliottrosenberg and @NoureldinYosri if you could take a look at this. Thanks!

eliottrosenberg reviewed (changes requested):
Very nice work Danni!! This looks really great!

A few initial comments:

Regarding line 55 in `shuffle_circuits_with_readout_benchmarking.py`, it should be possible to specify 0 for `num_random_bitstrings` in order to do the measurement without readout mitigation (or have a different way of running without readout mitigation).

I think you went with lists of tuples for `circuits_to_pauli` because you can't put a `cirq.Circuit` into a dictionary, but you can put a [`cirq.FrozenCircuit`](https://quantumai.google/reference/python/cirq/FrozenCircuit) into a dictionary, so maybe that would be more convenient (not sure). Also, the object that `measure_pauli_strings` outputs is pretty complicated, with lots of nested tuples, so maybe we want to create some data class for the output?

ddddddanni commented:
> Very nice work Danni!! This looks really great!
> 
> A few initial comments:
> 
> Regarding line 55 in `shuffle_circuits_with_readout_benchmarking.py`, it should be possible to specify 0 for `num_random_bitstrings` in order to do the measurement without readout mitigation (or have a different way of running without readout mitigation).
> 
> I think you went with lists of tuples for `circuits_to_pauli` because you can't put a `cirq.Circuit` into a dictionary, but you can put a [`cirq.FrozenCircuit`](https://quantumai.google/reference/python/cirq/FrozenCircuit) into a dictionary, so maybe that would be more convenient (not sure). Also, the object that `measure_pauli_strings` outputs is pretty complicated, with lots of nested tuples, so maybe we want to create some data class for the output?

Done. 
In the new commit, I allow ```shuffle_circuits_with_readout_benchmarking``` to take 0 for ```num_random_bitstrings```, and add a test to cover the situation. The design is the run_shuffled_with_readout_benchmarking method will return a empty SingleQubitReadoutCalibrationResult if no calibration is actually done.
I also allow measure_pauli_strings to take num_random_bitstrings = 0. In this case, no mitigation is actually done, and the mitigated result and unmitigated result are the same. Also add a test to handle this situation. (not sure if we really want to have this situation supported?)
Besides, I make the return type of measure_pauli_strings a data class ```CircuitToPauliStringsMeasurementResult```. Thanks for the suggestion, the code is more readable with this change.

eliottrosenberg reviewed (changes requested):
Thanks, Danni!! I tried testing it in [this](https://colab.research.google.com/drive/1nHzxUQuSIWyFtMynZmAexxgNv5AVjVct?usp=sharing) colab and got the error shown there.

ddddddanni commented:
> Thanks, Danni!! I tried testing it in [this](https://colab.research.google.com/drive/1nHzxUQuSIWyFtMynZmAexxgNv5AVjVct?usp=sharing) colab and got the error shown there.

Can you try again? The problem should be fixed in the latest commit ;)

eliottrosenberg reviewed (changes requested):
Thanks, Danni! I tested it again, and there seems to be a bug in the way it is computing the expectation value. I tried it on a simulator without noise in a case where the expectation value should be 1 and got something close to 0. https://colab.sandbox.google.com/drive/1_on4xIHQNMH_2km3RjlfoZxBcYim9A3y

Let me know if you have any trouble identifying the cause of this and I can step in and help.

eliottrosenberg commented:
@ddddddanni Here is a simpler example that illustrates the bug: https://colab.sandbox.google.com/drive/1DUOEbrJSHIvlIN1CI7T6lug4zGDB5nyG

NoureldinYosri commented on cirq-core/cirq/contrib/paulistring/pauli_string_measurement_with_readout_mitigation.py:
```diff
@@ -0,0 +1,312 @@
+from cirq.study import ResultDict
+
+
+@dataclasses.dataclass
```
prefer to use `attrs` over `dataclasses`

ddddddanni replied:
Done!

NoureldinYosri commented on cirq-core/cirq/contrib/paulistring/pauli_string_measurement_with_readout_mitigation.py:
```diff
@@ -0,0 +1,312 @@
+        elif pauli_op == ops.Y:
+            # Rotate to Y basis: Rx(pi/2)
+            operations.append(ops.rx(np.pi / 2)(qid_list[qubit_index]))
+        # No operation needed for Pauli Z or I (identity)
```
there shouldn't be identity in this case

ddddddanni replied:
Done!

NoureldinYosri commented on cirq-core/cirq/contrib/paulistring/pauli_string_measurement_with_readout_mitigation.py:
```diff
@@ -0,0 +1,312 @@
+    # Extract unique qubit tuples from all circuits
+    unique_qubit_tuples = set()
+    for circuit in circuits_to_pauli.keys():
+        unique_qubit_tuples.add(tuple(sorted(set(circuit.all_qubits()))))
```
`circuit.all_qubits()` returns a frozenset, so you can dop the `set` here and elsewhere

NoureldinYosri reviewed (commented):
LGTM

ddddddanni commented:
> @ddddddanni Here is a simpler example that illustrates the bug: https://colab.sandbox.google.com/drive/1DUOEbrJSHIvlIN1CI7T6lug4zGDB5nyG

Done! Can you try test again? The issue was Pauli I was incorrectly treated as Z in expectation calculation, and now the issue is fixed, and I also modified tests to test this.

eliottrosenberg reviewed (changes requested):
Thanks, @ddddddanni! This is really great! One more request: can we multiply the result of measuring the PauliString by its coefficient? For example, right now your code does not distinguish between the following PauliStrings:
```python
cirq.PauliString(cirq.Z.on_each(cirq.LineQubit.range(2)))
```
and
```python
-1 * cirq.PauliString(cirq.Z.on_each(cirq.LineQubit.range(2)))
```

eliottrosenberg commented:
(And when you implement my previous suggestion, make sure to also scale the statistical uncertainty by `abs(coefficient)`)

ddddddanni commented:
> Thanks, @ddddddanni! This is really great! One more request: can we multiply the result of measuring the PauliString by its coefficient? For example, right now your code does not distinguish between the following PauliStrings:
> 
> ```python
> cirq.PauliString(cirq.Z.on_each(cirq.LineQubit.range(2)))
> ```
> 
> and
> 
> ```python
> -1 * cirq.PauliString(cirq.Z.on_each(cirq.LineQubit.range(2)))
> ```

Sure! I added some codes to multiple the coefficient with the result of measuring the paulistring. Also added some more codes to validate the input and tests.

eliottrosenberg reviewed (approved):
Thanks, @ddddddanni, this is great!

ddddddanni commented:
> @ddddddanni LGTM ... I will take another look after you fix the failing CIs

Thanks! Failing checks should be fixed now ;)

ddddddanni commented on cirq-core/cirq/contrib/paulistring/pauli_string_measurement_with_readout_mitigation_test.py:
```diff
@@ -0,0 +1,565 @@
+    circuit_3 = cirq.FrozenCircuit(_create_ghz(8, qubits_3))
+
+    circuits_to_pauli: Dict[cirq.FrozenCircuit, list[cirq.PauliString]] = {}
+    circuits_to_pauli[circuit_1] = [
```
Done!

ddddddanni commented on cirq-core/cirq/contrib/shuffle_circuits/shuffle_circuits_with_readout_benchmarking_test.py:
```diff
@@ -170,14 +175,141 @@ def test_shuffled_circuits_with_readout_benchmarking_errors_with_noise_and_input
+    readout_qubits = [qubits_1, qubits_2]
+
+    # Generate random input circuits and append measurements
+    input_circuit_1 = rqcg.generate_library_of_2q_circuits(
```
Done!

ddddddanni commented on cirq-core/cirq/contrib/paulistring/pauli_string_measurement_with_readout_mitigation.py:
```diff
@@ -0,0 +1,387 @@
+def measure_pauli_strings(
+    circuits_to_pauli: Dict[circuits.FrozenCircuit, list[ops.PauliString]],
+    sampler: work.Sampler,
+    rng_or_seed: Union[np.random.Generator, int],
```
Done!

ddddddanni commented on cirq-core/cirq/contrib/paulistring/pauli_string_measurement_with_readout_mitigation.py:
```diff
@@ -0,0 +1,387 @@
+
+    Returns:
+        A list of PauliStringMeasurementResult objects, where each object contains:
+            - The Pauli string that was measured.
```
Sure! Changed.

ddddddanni commented on cirq-core/cirq/contrib/paulistring/pauli_string_measurement_with_readout_mitigation.py:
```diff
@@ -0,0 +1,387 @@
+    """Builds a list of empty confusion matrices"""
+    cms: list[np.ndarray] = []
+    for _ in range(qubits_length):
+        cms.append(_build_one_qubit_confusion_matrix(0, 0))
```
Sure! I changed the method to directly return the [_build_one_qubit_confusion_matrix(0, 0) for _ in range(qubits_length)]

ddddddanni commented on cirq-core/cirq/contrib/paulistring/pauli_string_measurement_with_readout_mitigation.py:
```diff
@@ -0,0 +1,387 @@
+        if qubit in pauli_string:
+            pauli_op = pauli_string[qubit]
+            if pauli_op == ops.X:
+                operations.append(ops.ry(-np.pi / 2)(qubit))
```
Done!

ddddddanni commented on cirq-core/cirq/contrib/paulistring/pauli_string_measurement_with_readout_mitigation.py:
```diff
@@ -0,0 +1,387 @@
+
+def _validate_input(
+    circuits_to_pauli: Dict[circuits.FrozenCircuit, list[ops.PauliString]],
+    rng_or_seed: Union[np.random.Generator, int],
```
Sure! Moved this to the end.

ddddddanni commented on cirq-core/cirq/contrib/paulistring/pauli_string_measurement_with_readout_mitigation.py:
```diff
@@ -0,0 +1,387 @@
+    calibration_result: Optional[SingleQubitReadoutCalibrationResult]
+
+
+@attrs.define
```
Done!

ddddddanni commented on cirq-core/cirq/contrib/paulistring/pauli_string_measurement_with_readout_mitigation.py:
```diff
@@ -0,0 +1,387 @@
+    mitigated_stddev: float
+    unmitigated_expectation: float
+    unmitigated_stddev: float
+    calibration_result: Optional[SingleQubitReadoutCalibrationResult]
```
Done!

ddddddanni commented on cirq-core/cirq/contrib/paulistring/pauli_string_measurement_with_readout_mitigation.py:
```diff
@@ -0,0 +1,387 @@
+from cirq.study import ResultDict
+
+
+@attrs.define
```
Sure, changed it to @attrs.frozen

NoureldinYosri commented:
great work @ddddddanni ... please fix the coverage CI and then we can merge this PR

ddddddanni commented:
> great work @ddddddanni ... please fix the coverage CI and then we can merge this PR

Thanks!! The coverage check should be fixed now ; )
